docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+
+Since object keys now include a generation suffix, the index of these keys must also be updated.
+
+IndexPart currently stores keys and LSNs sufficient to reconstruct key names: this would be
+extended to store the generation numbers with which a particular object was written, to enable
+reconstructing the suffix as well.
+
+This will increase the size of the file, but only modestly: layers are already encoded as
+their string-ized form, so the overhead is about 20 bytes per layer. This will be less if/when
+the index storage format is migrated to a binary format from JSON.
+
+### Pageserver startup changes
+
+- The pageserver must obtain a generation number by some means (see Control Plane Changes below) before
+  doing any remote writes.
+- The pageserver _may_ also do some synchronization of its attachments to see which are still
+  valid, see "Synchronizing attachments on pageserver startup" in the optimizations section.
+
+### `pageserver::Tenant` startup changes
+
+#### Attachment
+
+We already reconcile with remote metadata during attachment. It is necessary to synchronize
+with remote metadata even if there is some local metadata that could be brought up to
+date by ingesting the WAL, because if another node updated
+the remote storage, it might have advanced the remote_consistent_lsn such that the safekeeper drops WAL
+content that would be needed for this node to recover. Reconcilation with remote metadata also acts as an optimization, to avoid redundantly
+re-ingesting WAL data that is already in S3.
+
+Because index files are now suffixed with generation numbers, some changes are needed
+to load them:
+
+- In a clean migration of a tenant, the control plane may remember the tenant's most recent
+  generation numbers, and provide them to the new node in the attach request. This is sufficient
+  for the new node to directly GET the latest index.
+- Alternatively, we may use the storage broker for this: if we are continuously publishing the generation
+  information to the broker, then a newly attached tenant may query this information back and if
+  the attachment generation is exactly 1 less than the current attachment, or if the attachment is
+  the same but the node generation is exactly 1 less and the node ID is the same, then we may use that
+  information to calculate the most recent index key. See "Optional Safekeeper changes" below for where
+  we publish generation info to the safekeeper & in turn to the storage broker.
+- As a fallback, newly attached tenants may issue a ListObjectsv2 request using index_part as a prefix
+  to enumerate the indices and pick the most recent one by attachment generation. The listing would
+  typically only return 1-2 indices.
+
+Once the new attachment has written out its index_part.json, it may clean up historic index_part.json
+files that were found, unless the control plane has indicated to us that the tenant is multiply attached
+(see the subsequent HA RFC for this concept).
```

Comment:
In general, who cleans up historic index parts? Are they on the deletion queue?

(In my MVCC proposal, they were part of the MVCC scheme.)

Reply:
We have some flexibility here, as deletion of old indices is essentially an optimization.

I'm imagining that we kick off some background task after the first time we write a remote index, and then it basically just does a ListObjectsv2 with the index prefix and then submits those to the deletion queue.  But in terms of code structure it might be simpler to just let the scrub code handle this: treat all non-current-generation index files as garbage.  Updated this section to be clearer and describe the options for who actually doers it. 868dc1d7